File: backend/internal/account/account.go
// Пакет account — учётные записи, места в гараже и их баланс проверок.
//
// Устройство баланса. Проверки лежат не на аккаунте, а на месте гаража, то
// есть у конкретной машины: купленные для одной машины не тратятся на другую.
// Это решение продуктовое, но оно определяет всю модель, поэтому здесь нет
// общего счёта — только счёт места.
//
// Машина запирается на месте после первой потраченной проверки. До неё
// правки свободны: опечатка в годе выпуска не должна стоить человеку места.
// После — марку, модель, поколение и год менять нельзя, иначе одно место
// обслуживало бы сколько угодно машин по кругу. Пробег правится всегда:
// он растёт, и от него зависит разбор.
package account

import (
	"errors"
	"fmt"
	"time"

	"stuk/backend/internal/billing"
)

var (
	ErrNoSlot         = errors.New("места с таким номером нет")
	ErrSlotLocked     = errors.New("машину уже разбирали: менять и удалять её нельзя")
	ErrSlotTaken      = errors.New("место занято")
	ErrSlotEmpty      = errors.New("на месте нет машины")
	ErrNoChecks       = errors.New("на этой машине не осталось проверок")
	ErrUnknownProduct = errors.New("неизвестный товар")
)

// Car — машина на месте. Пробег отделён от остальных полей намеренно:
// он единственный меняется в течение жизни машины.
type Car struct {
	Make       string `json:"make"`
	Model      string `json:"model"`
	Generation string `json:"generation,omitempty"`
	Year       int    `json:"year"`
	Mileage    int    `json:"mileage"`
}

// SameIdentity сравнивает то, что делает место занятым именно этой машиной.
// Пробег в сравнение не входит.
func (c Car) SameIdentity(o Car) bool {
	return c.Make == o.Make && c.Model == o.Model &&
		c.Generation == o.Generation && c.Year == o.Year
}

// Slot — место в гараже со своим балансом.
type Slot struct {
	ID     string `json:"id"`
	Car    *Car   `json:"car,omitempty"`
	Checks int    `json:"checks"`
	// Сколько проверок на этом месте уже потрачено. Ненулевое значение
	// запирает машину: именно потраченная проверка, а не сам факт покупки.
	Used int `json:"used"`
}

// Locked — машину менять и удалять нельзя.
func (s Slot) Locked() bool { return s.Used > 0 }

// Pending — купленные проверки, которые ещё не легли на место.
//
// Покупка засчитывается сервером сразу, а привязка может не дойти: связь
// рвётся ровно в тот момент, когда деньги уже списаны. Такие проверки ждут
// здесь, и приложение привязывает их позже, ничего не потеряв.
type Pending struct {
	PurchaseID string `json:"purchase_id"`
	Checks     int    `json:"checks"`
	// Место, для которого покупали. Пустое — покупатель не выбрал или
	// место успели удалить.
	SlotID string `json:"slot_id,omitempty"`
}

// Account — учётная запись, привязанная к аккаунту Google или Apple.
type Account struct {
	ID       string `json:"id"`
	Provider string `json:"provider"`
	Subject  string `json:"subject"`
	Created  string `json:"created"`
	// Растёт при каждом изменении. Приложение держит копию состояния у себя
	// и по версии понимает, что она устарела, не выкачивая всё заново.
	Version int64 `json:"version"`

	Slots   []Slot    `json:"slots"`
	Pending []Pending `json:"pending,omitempty"`

	// Обработанные покупки: магазин присылает уведомление повторно, и без
	// этого списка один чек начислялся бы много раз.
	Done map[string]bool `json:"done,omitempty"`
}

// New создаёт запись с бесплатным местом и проверками на нём.
func New(id, provider, subject string, now time.Time) *Account {
	a := &Account{
		ID:       id,
		Provider: provider,
		Subject:  subject,
		Created:  now.UTC().Format(time.RFC3339),
		Done:     map[string]bool{},
	}
	for i := 0; i < billing.FreeSlots; i++ {
		a.Slots = append(a.Slots, Slot{
			ID:     fmt.Sprintf("s%d", i+1),
			Checks: billing.FreeChecks,
		})
	}
	return a
}

func (a *Account) slot(id string) *Slot {
	for i := range a.Slots {
		if a.Slots[i].ID == id {
			return &a.Slots[i]
		}
	}
	return nil
}

// SetCar ставит машину на пустое место.
func (a *Account) SetCar(slotID string, car Car) error {
	s := a.slot(slotID)
	if s == nil {
		return ErrNoSlot
	}
	if s.Car != nil {
		return ErrSlotTaken
	}
	s.Car = &car
	return nil
}

// EditCar меняет машину на месте. После первой проверки проходит только
// правка пробега — остальное запёрто.
func (a *Account) EditCar(slotID string, car Car) error {
	s := a.slot(slotID)
	if s == nil {
		return ErrNoSlot
	}
	if s.Car == nil {
		return ErrSlotEmpty
	}
	if s.Locked() && !s.Car.SameIdentity(car) {
		return ErrSlotLocked
	}
	s.Car = &car
	return nil
}

// RemoveCar освобождает место. Запертое место не освобождается: иначе запрет
// на смену машины обходился бы удалением и добавлением заново.
func (a *Account) RemoveCar(slotID string) error {
	s := a.slot(slotID)
	if s == nil {
		return ErrNoSlot
	}
	if s.Locked() {
		return ErrSlotLocked
	}
	s.Car = nil
	return nil
}

// Spend списывает проверку с места. Возвращает ошибку, если проверок нет —
// в этом случае разбор не начинается и деньги не тратятся.
func (a *Account) Spend(slotID string) error {
	s := a.slot(slotID)
	if s == nil {
		return ErrNoSlot
	}
	if s.Car == nil {
		return ErrSlotEmpty
	}
	if s.Checks <= 0 {
		return ErrNoChecks
	}
	s.Checks--
	s.Used++
	return nil
}

// Refund возвращает проверку, если разбор сорвался не по вине человека.
// Счётчик потраченных не откатывается: машина уже отправлялась в разбор,
// и отпирать её обратно нельзя — иначе отмена станет способом обхода.
func (a *Account) Refund(slotID string) {
	if s := a.slot(slotID); s != nil {
		s.Checks++
	}
}

// Grant начисляет покупку. Повторный вызов с тем же чеком ничего не меняет:
// магазины присылают уведомления не по одному разу.
//
// slotID — место, для которого покупали проверки. Если места нет или оно
// не указано, проверки уходят в Pending и ждут привязки.
func (a *Account) Grant(purchaseID, productID, slotID string) error {
	p, ok := billing.ByID(productID)
	if !ok {
		return ErrUnknownProduct
	}
	if a.Done == nil {
		a.Done = map[string]bool{}
	}
	if a.Done[purchaseID] {
		return nil
	}
	a.Done[purchaseID] = true

	for i := 0; i < p.Slots; i++ {
		a.Slots = append(a.Slots, Slot{
			ID:     fmt.Sprintf("s%d", len(a.Slots)+1),
			Checks: p.ChecksPerSlot,
		})
	}
	if p.Checks > 0 {
		if s := a.slot(slotID); s != nil {
			s.Checks += p.Checks
		} else {
			a.Pending = append(a.Pending, Pending{
				PurchaseID: purchaseID, Checks: p.Checks, SlotID: slotID,
			})
		}
	}
	return nil
}

// Assign кладёт отложенные проверки на место.
func (a *Account) Assign(purchaseID, slotID string) error {
	s := a.slot(slotID)
	if s == nil {
		return ErrNoSlot
	}
	for i, p := range a.Pending {
		if p.PurchaseID != purchaseID {
			continue
		}
		s.Checks += p.Checks
		a.Pending = append(a.Pending[:i], a.Pending[i+1:]...)
		return nil
	}
	return fmt.Errorf("покупка %q не найдена среди неразнесённых", purchaseID)
}

File: app/lib/data/account.dart
/// Учётная запись и её гараж — так, как их видит сервер.
///
/// Сервер здесь главный: баланс проверок и замок на машине решает он.
/// Приложение держит копию, чтобы показывать гараж мгновенно и без сети,
/// но ничего по ней не разрешает: попытка разбора всё равно уйдёт на сервер
/// и там будет принята или отклонена.
library;

class Car {
  final String make;
  final String model;
  final String generation;
  final int year;
  final int mileage;

  const Car({
    required this.make,
    required this.model,
    this.generation = '',
    required this.year,
    required this.mileage,
  });

  /// Пробег в сравнение не входит: он меняется всю жизнь машины, а остальное
  /// после первого разбора заперто.
  bool sameIdentity(Car o) =>
      make == o.make &&
      model == o.model &&
      generation == o.generation &&
      year == o.year;

  factory Car.fromJson(Map<String, dynamic> j) => Car(
        make: j['make'] as String? ?? '',
        model: j['model'] as String? ?? '',
        generation: j['generation'] as String? ?? '',
        year: (j['year'] as num?)?.toInt() ?? 0,
        mileage: (j['mileage'] as num?)?.toInt() ?? 0,
      );

  Map<String, dynamic> toJson() => {
        'make': make,
        'model': model,
        if (generation.isNotEmpty) 'generation': generation,
        'year': year,
        'mileage': mileage,
      };
}

/// Место в гараже со своим балансом проверок.
class Slot {
  final String id;
  final Car? car;
  final int checks;

  /// Сколько проверок уже потрачено. Ненулевое значение запирает машину:
  /// менять и удалять её нельзя.
  final int used;

  const Slot({required this.id, this.car, required this.checks, this.used = 0});

  bool get locked => used > 0;
  bool get isEmpty => car == null;

  factory Slot.fromJson(Map<String, dynamic> j) => Slot(
        id: j['id'] as String? ?? '',
        car: j['car'] == null
            ? null
            : Car.fromJson((j['car'] as Map).cast<String, dynamic>()),
        checks: (j['checks'] as num?)?.toInt() ?? 0,
        used: (j['used'] as num?)?.toInt() ?? 0,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        if (car != null) 'car': car!.toJson(),
        'checks': checks,
        'used': used,
      };
}

/// Купленные проверки, которые ещё не легли на место: связь могла оборваться
/// сразу после оплаты. Приложение доносит их до машины позже.
class PendingChecks {
  final String purchaseId;
  final int checks;
  final String slotId;

  const PendingChecks({
    required this.purchaseId,
    required this.checks,
    this.slotId = '',
  });

  factory PendingChecks.fromJson(Map<String, dynamic> j) => PendingChecks(
        purchaseId: j['purchase_id'] as String? ?? '',
        checks: (j['checks'] as num?)?.toInt() ?? 0,
        slotId: j['slot_id'] as String? ?? '',
      );

  Map<String, dynamic> toJson() => {
        'purchase_id': purchaseId,
        'checks': checks,
        if (slotId.isNotEmpty) 'slot_id': slotId,
      };
}

class AccountState {
  final String id;
  final String provider;
  final List<Slot> slots;
  final List<PendingChecks> pending;

  /// Растёт при каждом изменении на сервере. По нему приложение спрашивает
  /// «у меня всё ещё свежее?» одним коротким запросом вместо выкачивания.
  final int version;

  const AccountState({
    required this.id,
    this.provider = '',
    this.slots = const [],
    this.pending = const [],
    this.version = 0,
  });

  Slot? slot(String id) {
    for (final s in slots) {
      if (s.id == id) return s;
    }
    return null;
  }

  /// Машины, стоящие на местах. Пустые места — это ёмкость гаража, а не машины.
  List<Slot> get occupied => [for (final s in slots) if (!s.isEmpty) s];

  bool get hasFreeSlot => slots.any((s) => s.isEmpty);

  factory AccountState.fromJson(Map<String, dynamic> j) => AccountState(
        id: j['id'] as String? ?? '',
        provider: j['provider'] as String? ?? '',
        slots: [
          for (final s in (j['slots'] as List? ?? []))
            Slot.fromJson((s as Map).cast<String, dynamic>())
        ],
        pending: [
          for (final p in (j['pending'] as List? ?? []))
            PendingChecks.fromJson((p as Map).cast<String, dynamic>())
        ],
        version: (j['version'] as num?)?.toInt() ?? 0,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'provider': provider,
        'slots': [for (final s in slots) s.toJson()],
        'pending': [for (final p in pending) p.toJson()],
        'version': version,
      };
}
